### ISSUE TITLE
Wrong directory structure after unpacking if zip contains only a directory

### ISSUE THREAD

**Issue by BreiteSeite** [composer/composer#3945]
Hi,

we have the following structure in some of our Zips:
```
package/
    src/
        PhpFile1.php
        PhpFile2.php
```

However, after installing this package the directory structure suddenly differes from the one in the archive:
```
src/
    PhpFile1.php
    PhpFile2.php
```

We identified the code which causes this (wrong) behavior at [ArchiveDownloader.php](https://github.com/composer/composer/blob/7912253df619a924e7091a022d20a77ad2d3939b/src/Composer/Downloader/ArchiveDownloader.php#L56-L59).

Is there any reason for this behavior? I think it clearly is a bug.

**Comment by slbmeh** [comment]
I believe it is intended to work exactly as it does in this scenario. There are two very common competing archive formats:

```
package/
    src/
        file1
    test/
        file2
````

vs

```
src/
    file1
test/
    file2
```

The `composer.json` needs to be in the root, that code allows both styles of archiving to coexist. If you intend for your archive to keep that structure you need to have your `composer.json` in the root, which in that case it wouldn't have only a single directory and would not change your structure.

**Comment by BreiteSeite** [comment]
> The composer.json needs to be in the root

This assumption is wrong. We exclude it from the archive because we just include the source code needed for running the package. Which results in only 1 top-level directory without files. However, this behavior of composer **breaks the autoloading** for these packages.

> you need to have your composer.json in the root, which in that case it wouldn't have only a single directory and would not change your structure.

We know that this is a workaround for this bug, however, it still is a bug.

**Comment by alcohol** [comment]
>> The composer.json needs to be in the root

> This assumption is wrong.

It's not an assumption...

**Comment by BreiteSeite** [comment]
>> > The composer.json needs to be in the root

>> This assumption is wrong.

> It's not an assumption...

Fine. The "requirement" is wrong. Also: it isn't documented.

**Comment by alcohol** [comment]
But if your packages contain no composer.json, then how do you even define your autoload definitions for these packages?

**Comment by BreiteSeite** [comment]
The autoload information is gathered via satis/packagist and written to composer.lock, therefore there is no need for the composer.json/lock file inside the archive.

**Comment by alcohol** [comment]
![wat](https://memecomiccentral.files.wordpress.com/2012/02/meme-face-2.jpg)

**Comment by slbmeh** [comment]
> > you need to have your composer.json in the root, which in that case it wouldn't have only a single directory and would not change your structure.

> We know that this is a workaround for this bug, however, it still is a bug.

It isn't a workaround.

> > > > The composer.json needs to be in the root

> > > This assumption is wrong.

> > It's not an assumption...

> Fine. The "requirement" is wrong. Also: it isn't documented.

It is explicitly documented: https://getcomposer.org/doc/05-repositories.md#artifact 
> Each zip artifact is just a ZIP archive with composer.json in root folder

Packagist doesn't work with artifact repositories and Satis shouldn't generate artifacts like that. Composer generally doesn't work for a package without a `composer.json`, which is why it needs to be in the root of the project.  If you must have directories placed in between your path to your root directory and the vendor director that is what `target-dir` is for.

**Comment by BreiteSeite** [comment]
Okay, i think we are talking about 2 different things. While you refer to repository artifacts (= artifacts which ARE the repository), i talk about distribution artifacts (= artifacts which i get FROM/THROUGH a repository (in my case: satis)).

Therefore, the link you posted is irrelevant because it's about the former (repository artifacts). And the `target-dir` configuration is completely irrelevant for the issue.

Sorry about the confusion and for not describing the issue clearly enough in the first place.
Now that this is (hopefully) cleared, does the issue makes more sense?

**Comment by alcohol** [comment]
Lets take a step further back then.

How are these 'artifacts' you speak of defined in your root composer.json? Please share with us your composer.json so we can better understand your configuration. Because the way you are describing it right now we can't make heads or tails of this issue.

**Comment by BreiteSeite** [comment]
Given the following situation:

```
Application:
	- htdocs
		- index.php
	- vendor
		- vendor1
			- package1 (package1 contents)
				- package
					- src
						- ....
				- composer.json
				- composer.lock
	- composer.json
	- composer.lock
```

See we have the applications composer.json and the package1's composer.json. Given in package1's composer.json contains the following:

```json
"archive": {
	"exclude": [
		"composer.json",
		"composer.lock"
	]
}
```

The filesystem architecture which is except after `composer install` in the application directory is the following:

```
Application:
	- htdocs
		- index.php
	- vendor
		- vendor1
			- package1 (package1 contents)
				- package
					- src
						- ....
	- composer.json
	- composer.lock
```

However, it is the following:

```
Application:
	- htdocs
		- index.php
	- vendor
		- vendor1
			- package1 (package1 contents)
				- src (<- missing package)
					- ....
	- composer.json
	- composer.lock
```

Which breaks autoloading, because the autoloader tries to search for classes in vendor/vendor1/package1/**package**/src/....

**Comment by alcohol** [comment]
I seriously don't understand what you are doing. Where is the contents of your  `composer.json`?

**Comment by stof** [comment]
@alcohol Composer actually never reads the composer.json inside the downloaded archive. It already has all the metadata before that (from Packagist/Satis/Git/...).

@BreiteSeite my advice would be to keep a file at the root when building your archive (it can be the composer.json, or the readme, or even an empty ``.keep`` file).
The support of archives created with an extra root folder cannot be dropped from Composer: it would require dropping the support of Github archives, and github archives are the main source of downloads for composer (given that Packagist does not generate its own archives and that most packages registered on Packagist are hosted on github)

**Comment by alcohol** [comment]
I'm just confused.. if he has the packages with a composer.json, why repackage them as a zip only to install them again using the archive method? Why not just directly install them? And why do the composer.json and composer.lock absolutely have to be removed?

**Comment by BreiteSeite** [comment]
@stof thanks for participating in the discussion.

Yes, we workaround that issue already (with adding more files like README.md's etc to the archives). Talking about support for GitHub archives makes it so much clearer why this behavior is needed. Thanks.

However, could we improve the logic for doing this? For example:
* Only apply it, if the package was downloaded from github.com
or if this isn't possible:
* only apply it if the "single top-level-directory" looks like a directory created by github? They seem to name it like {repository-name}-{version}

Also: could you extend the comment from that source to include the reason that it is for github.com archives?

Thanks

**Comment by stof** [comment]
@BreiteSeite other tools could create archives the same way. It is not specific to github.com (I know github does it this way because this is the reason why the support was added in composer). I just checked and Bitbucket generates archives with a top-level directory too for instance.
So forcing any other service to provide archives without the top-level directory is not an option.

We could improve the Composer archiver to enforce having at least 2 items at the root though. In case there is a single folder left in the archive after applying the exclusion rules, it could add an empty ``.keep`` file at the root of the archive. This would solve the issue for any archive generated by the composer archiver (for instance in Satis). What do you think about it ?

@alcohol he does not have the package already with composer.json. Composer already has only the package **metadata** aggregated by Packagist/Satis to perform the dependency resolution. The **code** still need to be downloaded. And the goal of @BreiteSeite when excluding files from the archives is to make the archives smaller by keeping only the necessary files.

**Comment by BreiteSeite** [comment]
@stof Ah okay. I still think it's better to whitelist the providers which do that (i think there shouldn't be that many?), because modifying the archive to include a additional directory-layer is the real issue in the first way (in my opinion).
But i can understand that you don't want to risk breaking some hosts.

However, enforcing the composer archiver to have more than 1 top-level directory/file in the archive sounds like a "good enough" workaround. So we work around a workaround. :-|
Would it probably be better to just abort the archiving with an exception instead of modifying the archive? I don't know if silently adding something to the archive may break some 3rd-party validating/checksumming tools or cause other unexpected behavior. The downside would be that we break satis even if it could create a "working" archive.

Long story short: I could easily live with the workaround you suggested.

**Comment by alcohol** [comment]
How about

* if the archive contains a `composer.json`, consider its parent directory the root
* else keep archive structure intact

?

**Comment by stof** [comment]
@alcohol this would also be a BC break: Github relies on the git archive tool, which can be configured. This means that people could exclude the composer.json from the github archives (even though I have never seen it used for excluding  composer.json yet, but only for tests, doc and so on) and in this case, things would break. Other hosting providers may rely on ``git archive`` too btw

@BreiteSeite a whitelist is not an option. Gitlab also generates such archives. And given that it is meant to be self-hosted (additionally to the gitlab.com hosted instance), there is no way for Composer to build such whitelist.
And I don't agree with you saying it is not that common. All code hosting I tried until now are using the top level directory when generating their archive. It is a very common way (the goal being to make things easier for people to unzip it without messing their current working directory by keeping everything in a single folder).

Another way to solve the issue would be to use a top-level directory when generating archives with Composer.

**Comment by BreiteSeite** [comment]
> Another way to solve the issue would be to use a top-level directory when generating archives with Composer.

Nice idea. Sounds good to me.

**Comment by alcohol** [comment]
@stof ok, understandable. If keeping BC is important then maybe this expected archive format should be documented somewhere more clearly to avoid questions in the future.

**Comment by Seldaek** [comment]
I think it's best to keep this as is for consistency and BC.. If someone wants to send a doc PR go ahead.
